src/aspire/orientation/commonline_sync.py

```diff
@@ -1,34 +1,298 @@
+
+        for i in range(k - 1):
+            stmp = np.zeros((2, 2, k))
+            for j in range(i + 1, k):
+                stmp[:, :, j] = self._syncmatrix_ij_vote(
+                    clmatrix, i, j, np.arange(k), l)
+
+            for j in range(i + 1, k):
+                r22 = stmp[:, :, j]
+                s[2 * i:2 * (i + 1), 2 * j:2 * (j + 1)] = r22
+                s[2 * j:2 * (j + 1), 2 * i:2 * (i + 1)] = r22.T
+
+        self.syncmatrix = s
+
+    def _syncmatrix_ij_vote(self, clmatrix, i, j, k, l):
         """
-        constructor of an object for estimating 3D orientations
+        Compute the (i,j) rotation block of the synchronization matrix using voting method
         """
-        pass
+        tol = 1e-12
+
+        good_k, _, _ = self._vote_ij(clmatrix, l, i, j, k)
+
+        rs, good_rotations = self._rotratio_eulerangle_vec(
+            clmatrix, i, j, good_k, l)
+
+        if len(good_rotations) > 0:
+            rk = np.mean(rs, 2)
+            tmp_r = rs[:2, :2]
+            diff = tmp_r - rk[:2, :2, np.newaxis]
+            err = np.linalg.norm(diff) / np.linalg.norm(tmp_r)
+            if err > tol:
+                pass
+        else:
+            rk = np.zeros((3, 3))
 
-    def estimate(self):
+        r22 = rk[:2, :2]
+        return r22
+
+    def _rotratio_eulerangle_vec(self, cl, i, j, good_k, n_theta):
         """
-        perform estimation of orientations
+        Compute the rotation that takes image i to image j
+
+        Given a 3x3 common lines matrix, where the index of each common line
+        is between 1 and n_theta.
         """
-        pass
+        
+        r = np.zeros((3, 3, len(good_k)))
+        if i == j:
+            return 0, 0
+
+        tol = 1e-12
+
+        idx1 = cl[good_k, j] - cl[good_k, i]
+        idx2 = cl[j, good_k] - cl[j, i]
+        idx3 = cl[i, good_k] - cl[i, j]
+
+        a = np.cos(2 * np.pi * idx1 / n_theta)
+        b = np.cos(2 * np.pi * idx2 / n_theta)
+        c = np.cos(2 * np.pi * idx3 / n_theta)
+
+        cond = 1 + 2 * a * b * c - (np.square(a)
+                                    + np.square(b) + np.square(c))
+        too_small_idx = np.where(cond <= 1.0e-5)[0]
+        good_idx = np.where(cond > 1.0e-5)[0]
+
+        a = a[good_idx]
+        b = b[good_idx]
+        c = c[good_idx]
+        idx2 = idx2[good_idx]
+        idx3 = idx3[good_idx]
+        c_alpha = (a - b * c) / np.sqrt(1 - np.square(b)) / np.sqrt(1 - np.square(c))
+
+        ind1 = np.logical_or(idx3 > n_theta / 2 + tol,
+                             np.logical_and(idx3 < -tol, idx3 > -n_theta / 2))
+        ind2 = np.logical_or(idx2 > n_theta / 2 + tol,
+                             np.logical_and(idx2 < -tol, idx2 > -n_theta / 2))
+        c_alpha[np.logical_xor(ind1, ind2)] = -c_alpha[np.logical_xor(ind1, ind2)]
+
+        aa = cl[i, j] * 2 * np.pi / n_theta
+        bb = cl[j, i] * 2 * np.pi / n_theta
+        alpha = np.arccos(c_alpha)
+
+        ang1 = np.pi - bb
```

Comment:
The code here is to obtain the Euler angles and rotation matrices from the common lines instead of the common lines indices from the rotation matrices in the function `common_line_from_rots` used for estimating shifts. It is some kind of reverse operation.

Reply:
Done in the commit bd02850.